activerecord/lib/active_record/connection_adapters/mysql_adapter.rb

```diff
@@ -317,41 +308,10 @@ def initialize_type_map(m) # :nodoc:
         register_class_with_precision m, %r(time)i,     Fields::Time
       end
 
-      def exec_without_stmt(sql, name = 'SQL') # :nodoc:
-        # Some queries, like SHOW CREATE TABLE don't work through the prepared
-        # statement API. For those queries, we need to use this method. :'(
-        log(sql, name) do
-          result = @connection.query(sql)
-          affected_rows = @connection.affected_rows
-
-          if result
-            types = {}
-            fields = []
-            result.fetch_fields.each { |field|
-              field_name = field.name
-              fields << field_name
-
-              if field.decimals > 0
-                types[field_name] = Type::Decimal.new
-              else
-                types[field_name] = Fields.find_type field
-              end
-            }
-
-            result_set = ActiveRecord::Result.new(fields, result.to_a, types)
-            result.free
-          else
-            result_set = ActiveRecord::Result.new([], [])
-          end
-
-          [result_set, affected_rows]
-        end
-      end
-
       def execute_and_free(sql, name = nil) # :nodoc:
         result = execute(sql, name)
         ret = yield result
-        result.free
+        result.free if result
```

Comment:
@kamipo Do you specifically need mysql2 to provide Mysql2::Result#free, or would it be sufficient to allow Ruby GC to take care of the result set? I sort of imagine that this method was written for the original Mysql gem that required the explicit free.

That said, if it helps to land this into Rails sooner, then I'll be sure to provide a Mysql2::Result#free per your PR.